torch/csrc/autograd/record_function.h

```diff
@@ -78,25 +77,69 @@ struct TORCH_API RecordFunction {
 
  private:
   void processCallbacks();
+  void copyInputs(const at::ArrayRef<c10::IValue>& values) {
+    inputs_.insert(inputs_.end(), values.begin(), values.end());
+  }
 
   Function* fn_ = nullptr;
   StringView name_;
   int64_t sequence_nr_ = -1;
 
   RecordFunction* parent_ = nullptr;
 
-  GetPackedInputsCallback inputs_cb_ = nullptr;
-  mutable bool inputs_initialized_ = false;
-  // initialized lazily by inputs_cb_
-  mutable std::vector<c10::IValue> inputs_;
+  std::vector<c10::IValue> inputs_;
+  bool initialized_ = false;
 };
 
+#define _RECORD_FUNCTION(fn, inputs_case) \
+  torch::autograd::profiler::RecordFunction guard; \
+  switch (torch::autograd::profiler::recordFunctionState()) { \
+    case torch::autograd::profiler::RecordFunctionState::NAME_ONLY: { \
+      guard.before(fn); \
+      break; \
+    } \
+    case torch::autograd::profiler::RecordFunctionState::NAME_AND_INPUTS: { \
+      inputs_case \
+      break; \
+    } \
+    case torch::autograd::profiler::RecordFunctionState::NO_CALLBACKS: { \
+      break; \
+    } \
+  }
+
+#define RECORD_FUNCTION_WITH_STACK(fn, stack_ref) \
+  _RECORD_FUNCTION(fn, guard.before(fn, stack_ref);)
+
+#define RECORD_FUNCTION_WITH_INPUTS_SEQ(fn, seq, ...) \
+  _RECORD_FUNCTION(fn, \
+    torch::jit::Stack stack({__VA_ARGS__}); \
+    guard.before(fn, torch::jit::last(stack, stack.size()), seq); \
+  )
+
+#define RECORD_FUNCTION_WITH_INPUTS(fn, ...) \
+  RECORD_FUNCTION_WITH_INPUTS_SEQ(fn, -1, __VA_ARGS__)
+
+#define RECORD_FUNCTION_WITH_INPUT_RANGE(fn, start, end) \
+  _RECORD_FUNCTION(fn, \
+    torch::jit::Stack stack(start, end); \
+    guard.before(fn, torch::jit::last(stack, stack.size())); \
+  )
```

Comment:
I feel like many of those macros could be inline functions, and that would be _far_ more readable